rcl_action/test/rcl_action/test_action_server.cpp

```diff
@@ -0,0 +1,526 @@
+
+#include "rcl_action/action_server.h"
+
+#include "rcl/error_handling.h"
+#include "rcl/rcl.h"
+
+#include "test_msgs/action/fibonacci.h"
+
+TEST(TestActionServerInitFini, test_action_server_init_fini)
+{
+  rcl_ret_t ret = rcl_init(0, nullptr, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  rcl_node_t node = rcl_get_zero_initialized_node();
+  rcl_node_options_t node_options = rcl_node_get_default_options();
+  ret = rcl_node_init(&node, "test_action_server_node", "", &node_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  const rosidl_action_type_support_t * ts = ROSIDL_GET_ACTION_TYPE_SUPPORT(test_msgs, Fibonacci);
+  const rcl_action_server_options_t options = rcl_action_server_get_default_options();
+  const char * action_name = "test_action_server_name";
+  rcl_action_server_t action_server = rcl_action_get_zero_initialized_server();
+
+  // Initialize with a null action server
+  ret = rcl_action_server_init(nullptr, &node, ts, action_name, &options);
+  EXPECT_EQ(ret, RCL_RET_INVALID_ARGUMENT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Initialize with a null node
+  ret = rcl_action_server_init(&action_server, nullptr, ts, action_name, &options);
+  EXPECT_EQ(ret, RCL_RET_NODE_INVALID) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Initialize with an invalid node
+  rcl_node_t invalid_node = rcl_get_zero_initialized_node();
+  ret = rcl_action_server_init(&action_server, &invalid_node, ts, action_name, &options);
+  EXPECT_EQ(ret, RCL_RET_NODE_INVALID) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Initialize with a null typesupport
+  ret = rcl_action_server_init(&action_server, &node, nullptr, action_name, &options);
+  EXPECT_EQ(ret, RCL_RET_INVALID_ARGUMENT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Initialize with a null name
+  ret = rcl_action_server_init(&action_server, &node, ts, nullptr, &options);
+  EXPECT_EQ(ret, RCL_RET_INVALID_ARGUMENT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Initialize with a null options
+  ret = rcl_action_server_init(&action_server, &node, ts, action_name, nullptr);
+  EXPECT_EQ(ret, RCL_RET_INVALID_ARGUMENT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Initialize with valid arguments
+  ret = rcl_action_server_init(&action_server, &node, ts, action_name, &options);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+  // Try to initialize again
+  ret = rcl_action_server_init(&action_server, &node, ts, action_name, &options);
+  EXPECT_EQ(ret, RCL_RET_ALREADY_INIT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Try to initialize another action server with the same name
+  // FIXME(jacobperron): test failing
+  // rcl_action_server_t same_name_action_server = rcl_action_get_zero_initialized_server();
+  // ret = rcl_action_server_init(&same_name_action_server, &node, ts, action_name, &options);
```

Comment:
@jacobperron I presume we're aiming to catch this by means of invalid service name checks, but it doesn't seem like that outcome means the service name is repeated (see [here](https://github.com/ros2/rcl/blob/master/rcl/src/rcl/service.c#L117))?

Reply:
Oh, that's a bug in `rcl_action/names.cc`.